base/doc/clsguide.tex

```diff
@@ -1181,6 +1181,113 @@ \subsection{Querying localisation}
 where |#1| is the user input and the first argument to
 \cs{MakeUppercaseAux} takes two arguments, the locale and input text.
 
+\subsection{Extended and expandable references of properties}
+
+A property is something that \LaTeX{} can track, such as a page number, a section number or a name. Such properties
+can be labeled, their current value written to the \texttt{aux}-file and referenced in the next compilation, similar to the standard \cs{label}/\cs{ref} commands. 
+
+\begin{decl}
+  |\RecordProperties|\arg{label}\arg{list of properties} 
+\end{decl}
+This command writes the current or shipout value (depending on the declaration of the property) 
+of the properties in the list into the \texttt{aux}-file labeled by \meta{label}. The arguments can contain commands which are expanded. \meta{label} can expand to an arbitrary string (as long as it can safely be written to the \texttt{aux}-file) but note that the label names of \cs{label} and \cs{RecordProperties} share a singe namespace. That means that you get a \texttt{Label `A' multiply defined} warning with the following code:
+\begin{verbatim}
+\label{A}\RecordProperties{A}{page}
+\end{verbatim}
+
+
+\begin{decl}
+ |\RefProperty|\arg{label}\arg{property}
+\end{decl}
+This command allows to reference the value of the property \meta{property} labeled by \meta{label}. Differently to the standard \cs{ref} command the command is expandable and the value can for example---if it is a number---be used in an assignment.
+
+\begin{verbatim}
+\section{A section}
+\RecordProperties{mylabel}{pagenum,counter}
+\RefProperty{mylabel}{counter} % outputs section
+\setcounter{mycounter}{\RefProperty{mylabel}{pagenum}} 
+\end{verbatim}
+
+
+
+As \cs{RefProperty} is expandable it can not issue a rerun warning if a label is not found. If needed such a warning can be forced by the following command:
+\begin{decl}
+ |\RefUndefinedWarn|\arg{label}\arg{property}
+\end{decl}
+
+\LaTeX{} predefines a set of properties, this set contains also the properties stored by the 
+standard \cs{label} command:
+\begin{description}
+ \item[\texttt{abspage}] The absolute value of the current page: starts at $1$ and increases
+   monotonically at each shipout.
+ \item[page] The current page as given by \cs{thepage}: this may or may not
+   be a numerical value, depending on the current style. Contrast with
+   \texttt{abspage}. You get this value also  with the standard \cs{label}/\cs{pageref}.
+ \item[\texttt{pagenum}] The current page as arabic number. This is suitable for integer operations and
+   comparisions.
+ \item[\texttt{currentlabel}] The content of \cs{@currentlabel}. This is the value that
+   you get also with the standard \cs{label}/\cs{ref}. 
```

Comment:
Reading over the list of predefined properties, this strikes me as inconsistent. Either we should have currentlabel, currentcounter and currenttarget (because all three are produced by explicit commands in the document like headings or \refstepcounter or the like) or we should always drop the "current" and just call it label. 

I think there is something in making the predefined properties concise so my voe goes for call this one "label".

Reply:
Strong agreement with these not being in amy sense "current" when the item is retrieved! 

"label" has other problems, since it also used for the "logical label", with which this "label" has zero (or even negative:-) connection!

Reply:
They are "current" in my opinion: code in LaTeX sets them up and here is a "current" value for them at any time in the document and that is what is retrieved. And on top of it this is what LaTeX called them for the last 40 years (and nobody wondered or objected). My point is only that a shorter name would do as well and for consistency it should be either "current" on all of them or on none. And given that "on all of them" means a lot of  extra tokens for no good reasing in the code I'm for "none" ie also here. 

As to "label" being overloaded, I agree, but that doesn't mean it couldn't be used. I think it can and it will not cause practical problems, on the contrary inventing a really new name for somehting that is linked to `\@currentlabel` would cause more problems and removing the `\@currentlabel` interface on the code-level would be a non-starter.

Reply:
I changed the text and will change the actual code in the ltproperties-activechars branch if that is ok.